File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int  n = s.length();
        Map<Character, Integer> lastSeen = new HashMap<>();
        int left = 0 , maxlen = 0;
        
        for(int right = 0 ; right < n ; right++){
            char ch = s.charAt(right);

            if (lastSeen.containsKey(ch) && lastSeen.get(ch) >= left){
                left = lastSeen.get(ch) + 1;
            }

            lastSeen.put(ch, right);
            maxlen = Math.max(maxlen, right - left + 1);
        }

        return maxlen ;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        n = len(s)
        start = 0
        end = 0
        l = []
        maxl = 0

        while end < n :
            if s[end] in l:
                maxl = max(maxl , end - start)
                while s[end] in l:
                    l.remove(s[start])
                    start += 1
                
            else:
                l.append(s[end])
                end += 1
        
        
        return max(maxl, end - start)
